console: implement minimal `console.group()`

Trott opened the pull request:
Node.js exposes `console.group()` and `console.groupEnd()` via the
inspector. These functions have no apparent effect when called from
Node.js without the inspector. We cannot easily hide them when Node.js
is started without the inspector because we support opening the
inspector during runtime via `inspector.port()`.

Implement a minimal `console.group()`/`console.groupEnd()`. More
sophisticated implementations are possible, but they can be done in
userland and/or features can be added to this at a later time. (It lacks
the `label` argument to `console.group()` right now, for example. How to
handle `label`, or even whether to handle it,  may become
a bikeshed discussion. Landing a minimal implementation first avoids the
pitfall of that discussion or a similar discussion delaying the
implementation indefinitely.)

Refs: https://github.com/nodejs/node/issues/12675
Fixes: https://github.com/nodejs/node/issues/1716

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
console

TimothyGu commented on lib/console.js:
```diff
@@ -25,6 +25,9 @@ const errors = require('internal/errors');
 const kCounts = Symbol('counts');
 
+// Track amount of indentation required via `console.group()`.
+let groupIndent = 0;
```
The state should be per-Console.

Trott replied:
Fixed, thanks.

TimothyGu commented on lib/console.js:
```diff
@@ -214,6 +219,15 @@ Console.prototype.countReset = function countReset(label = 'default') {
 };
 
+Console.prototype.group = function group() {
+  groupIndent++;
```
This browsers this also prints whatever's passed to group:

![screenshot from 2017-08-18 14-48-56](https://user-images.githubusercontent.com/1538624/29447421-60329884-8424-11e7-9e70-8e420c9deb74.png)

IMO it should be something like:

```js
Console.prototype.group = function group(...args) {
  this.log(...args);
  this.groupIndent++;
};

Trott replied:
An argument passed to `console.group()` gets used as a label. It's not obvious to me what the best way to distinguish labels from the rest of the output is. However, omitting it entirely is probably not the way to go, so I'll add something like this.

Trott replied:
OK, done. (I still need to document it.)

Trott replied:
Documented.

TimothyGu commented on lib/console.js:
```diff
@@ -83,6 +86,8 @@ function createWriteErrorHandler(stream) {
 
 function write(ignoreErrors, stream, string, errorhandler) {
+  if (groupIndent > 0)
+    string = `${'\t'.repeat(groupIndent)}${string}`;
```
Two spaces instead of a tab? Would look more like `util.indent` output.

Trott replied:
That's how I did it at first but figured a tab would allow people to configure the indentation without us having to add configuration in our code. You raise a good point. Consistency is good,  I'll change it to two spaces.

Trott replied:
Changed, thanks.

lpinca commented on test/parallel/test-console-group.js:
```diff
@@ -0,0 +1,40 @@
+
+assert.strictEqual(stdout, expectedOut);
+assert.strictEqual(stderr, expectedErr);
+common.restoreStdout();
```
Missing `common.restoreStderr()`?

Trott replied:
Oops, yes, thanks, will fix that.

Trott replied:
Fixed, thanks.

jasnell commented:
I'm not overly in love with this approach, especially given that it's trivial to break the visual formatting using new lines. e.g. `console.group(); console.log('hello\nworld')' console.groupEnd()`. While I'd be quite happy to have an implementation on this, if we do something and get it wrong it would be a semver-major to fix it. Not sure what else to suggest tho

Trott commented:
> I'm not overly in love with this approach, especially given that it's trivial to break the visual formatting using new lines

I thought about that but figured that could be a patch-level fix or a minor-level feature enhancement at a later time if it was really necessary. I'm not sure how much this feature is going to get used, to be honest.

And I suspect starting with a full-blown implementation will be riskier than starting with a minimal implementation.

I'm OK with deciding that we don't actually need to implement `console.group()` and let it be a no-op and people can implement it in the ecosystem if they want it. I'd want there to be consensus, though, that we should close https://github.com/nodejs/node/issues/12675 and https://github.com/nodejs/node/issues/1716 (perhaps after documenting that it's a no-op).

Trott commented:
`.groupCollapsed()` added as an alias for `.group()`. I think this is ready for another look.

TimothyGu commented on lib/console.js:
```diff
@@ -25,6 +25,9 @@ const errors = require('internal/errors');
 const kCounts = Symbol('counts');
 
+// Track amount of indentation required via `console.group()`.
+const groupIndent = Symbol('groupIndent');
```
Nit: kGroupIndent for consistency with the kCounts above.

TimothyGu commented on lib/console.js:
```diff
@@ -64,6 +67,8 @@ function Console(stdout, stderr, ignoreErrors = true) {
     this[k] = this[k].bind(this);
   }
+
+  this[groupIndent] = '';
```
Ideally this should probably be "grouped" with kCounts above.

Trott commented:
Rebased, addressed @TimothyGu's comments, force-pushed.

This could use some more @nodejs/ctc review. Seems like the right thing to do to me, but I want to make sure we have consensus on that. Otherwise, I'll go the document-that-it-doesn't-do-anything route.

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9789/

Trott commented:
Landed in c40229a9b80736f1fdb31fac169b70a1d6af8669

BridgeAR commented:
Uh, seems like I looked a tiny bit to late at this. Right now if I read the code correct only the very first line of console.log is going to be indented. util.format will is going to add new lines in objects if something exceeds the line length though. So the grouping should apply for all lines and not only to the very first one and this is also how browsers behave.
Right now it would have to be done by checking for new lines in the output and replacing those with the new indentation lvl. If #14881 lands as it is this could actually be done without any hassle by passing a `indentationLvl` option to util.inspect (that is called by util.format depending on the input). The option is meant to be private would we might actually consider to expose it and document it?

Trott commented:
@BridgeAR I'm OK reverting this if you or anyone else think it's a terrible mistake to land it as is (just as long as we do it before this ends up in a release).

That said, I think this is better than the current situation, which is `console.group()` exists but doesn't actually do anything at all. However, I can see an argument that a no-op is better than a minimal implementation that might be perceived as faulty. 

If you think the implementation here can be made to work smoothly with `indentationLvl` once that lands, even better.

BridgeAR commented:
@jasnell new lines added by the user to console.log will actually not break the indentation because those are always escaped.

BridgeAR commented:
@Trott I think it is good to add the functionality itself and I think we could just open a new PR to fix the indentation before the next release. WDYT?

jasnell commented:
I for one would have preferred a bit more review but I'll leave it up to you on whether to open a revert.

Trott commented:
@BridgeAR I think @jasnell's point is that in this implementation, this code:

```js
console.log('not indented');
console.group();
console.log('indented\nshould also be indented');
```

...results in this output:

```console
not indented
  indented
should also be indented
```

...whereas the expected output should be:

```console
not indented
  indented
  should also be indented
```

That said, that shortcoming was a choice. I'll create a `known_issues` test for the problem and also the one you pointed out too.

For comparison the current output in Node.js 8.4.0:

```console
not indented
indented
should also be indented
```

Trott commented:
> @Trott I think it is good to add the functionality itself and I think we could just open a new PR to fix the indentation before the next release. WDYT?

I'll add `dont-land` labels that can be removed once your fixes/improvements land.

(And again, if anyone thinks this ought to be reverted instead, I'm OK with that too.)

BridgeAR commented:
@Trott right, I forgot that util.format does not escape the strings when not passed to util.inspect.

Trott commented:
https://github.com/nodejs/node/pull/14999 for the multiline fix. (Thanks, @BridgeAR, for the "here's how you should do it" info. :-D )

MylesBorins commented:
Is this something we want to land on v8.x?

Trott commented:
> Is this something we want to land on v8.x?

@MylesBorins Yes. I've removed the `dont-land` labels. Sorry about not doing that sooner.
